mono/metadata/w32process-win32.c

```diff
@@ -202,131 +199,141 @@ process_quote_path (const gchar *path)
 
-	*cmd = proc_start_info->arguments;
+	*cmd = proc_start_info->coophandle->arguments;
 
-	if (process_complete_path (mono_string_chars (proc_start_info->filename), &spath)) {
+	// FIXME There are excess utf8 <=> gunichar2 conversions here.
+	// We are either returning spath, or spath + " " + cmd.
+	// Just use gunichar2. Maybe move logic to C#.
+
+	if (process_complete_path (proc_start_info->filename, &spath)) {
 		/* Seems like our CreateProcess does not work as the windows one.
 		 * This hack is needed to deal with paths containing spaces */
-		if (*cmd) {
-			cmd_utf8 = mono_string_to_utf8_checked (*cmd, &mono_error);
-			if (!mono_error_set_pending_exception (&mono_error)) {
-				new_cmd = g_strdup_printf ("%s %s", spath, cmd_utf8);
-				*cmd = mono_string_new_wrapper (new_cmd);
-				g_free (cmd_utf8);
-				g_free (new_cmd);
-			} else {
-				*cmd = NULL;
-			}
+		if (!MONO_HANDLE_IS_NULL (*cmd)) {
+			cmd_utf8 = mono_string_handle_to_utf8 (*cmd, error);
+			goto_if_nok (error, error);
+			new_cmd = g_strdup_printf ("%s %s", spath, cmd_utf8);
+			*cmd = mono_string_new_utf8_len_handle (mono_domain_get (), new_cmd, strlen (new_cmd), error);
+			goto_if_nok (error, error);
 		}
 		else {
-			*cmd = mono_string_new_wrapper (spath);
+			*cmd = mono_string_new_utf8_len_handle (mono_domain_get (), spath, strlen (spath), error);
+			goto_if_nok (error, error);
 		}
-
-		g_free (spath);
 	}
 
-	return (*cmd != NULL) ? TRUE : FALSE;
+exit:
+	g_free (spath);
+	g_free (cmd_utf8);
+	g_free (new_cmd);
+	return !MONO_HANDLE_IS_NULL (*cmd);
+error:
+	*cmd = NULL_HANDLE_STRING;
+	goto exit;
 }
 
 MonoBoolean
-ves_icall_System_Diagnostics_Process_CreateProcess_internal (MonoW32ProcessStartInfo *proc_start_info, HANDLE stdin_handle,
-							     HANDLE stdout_handle, HANDLE stderr_handle, MonoW32ProcessInfo *process_info)
+ves_icall_System_Diagnostics_Process_CreateProcess_internal (MonoW32ProcessStartInfoHandle proc_start_info,
+	HANDLE stdin_handle, HANDLE stdout_handle, HANDLE stderr_handle, MonoW32ProcessInfo *process_info, MonoError *error)
 {
+	MonoCreateProcessCoop coop;
+	mono_createprocess_coop_init (&coop, proc_start_info, process_info);
+
 	gboolean ret;
-	gunichar2 *dir;
+	gunichar2 *dir = NULL;
 	STARTUPINFO startinfo={0};
 	PROCESS_INFORMATION procinfo;
 	gunichar2 *env_vars = NULL;
-	MonoString *cmd = NULL;
+	MonoStringHandle cmd = NULL_HANDLE_STRING;
 	guint32 creation_flags;
 
 	mono_process_init_startup_info (stdin_handle, stdout_handle, stderr_handle, &startinfo);
 
 	creation_flags = CREATE_UNICODE_ENVIRONMENT;
-	if (proc_start_info->create_no_window)
+	if (MONO_HANDLE_GETVAL (proc_start_info, create_no_window))
 		creation_flags |= CREATE_NO_WINDOW;
 	
-	if (process_get_shell_arguments (proc_start_info, &cmd) == FALSE) {
+	if (process_get_shell_arguments (coop, &cmd, error) == FALSE) {
 		process_info->pid = -ERROR_FILE_NOT_FOUND;
-		return FALSE;
+		ret = FALSE;
+		goto exit;
 	}
 
 	if (process_info->env_variables) {
```

Comment:
Is a raw pointer ok if just checking for null?

Reply:
Could the compiler cache it on the line that makes a handle out of it?

Reply:
I think this is okay.